File: src/tasks/cli/command/list/ListParser.java
package tasks.cli.command.list;

import static tasks.cli.parser.ParserUtil.assertNoExtraArgs;

import org.apache.commons.cli.CommandLine;
import tasks.cli.parser.ParserUtil;
import tasks.cli.parser.CommandParser;

/** Command line argument parser for the List command. */
public final class ListParser implements CommandParser<ListArguments> {

  @Override
  public ListArguments parse(CommandLine commandLine) {
    /*
     * No non-flag parameters allowed
     * optional --blocked flag
     * optional --completed flag
     * optional --all flag
     */
    assertNoExtraArgs(commandLine);

    boolean isAllSet = ParserUtil.getFlagPresence(commandLine, ListCommand.ALL_OPTION.value());
    boolean isBlockedSet =
        isAllSet || ParserUtil.getFlagPresence(commandLine, ListCommand.BLOCKED_OPTION.value());
    boolean isCompletedSet =
        isAllSet || ParserUtil.getFlagPresence(commandLine, ListCommand.COMPLETED_OPTION.value());
    boolean isUnblockedSet =
        isAllSet
            || ParserUtil.getFlagPresence(commandLine, ListCommand.UNBLOCKED_OPTION.value())
            || (!isBlockedSet && !isCompletedSet);

    return new ListArguments(isUnblockedSet, isBlockedSet, isCompletedSet);
  }
}

File: cli/src/me/strangepan/tasks/cli/feature/list/ListParser.kt
package me.strangepan.tasks.cli.feature.list

import org.apache.commons.cli.CommandLine
import me.strangepan.tasks.cli.parser.CommandParser
import me.strangepan.tasks.cli.parser.ParserUtil

/** Command line argument parser for the List command.  */
class ListParser : CommandParser<ListArguments> {
  override fun parse(commandLine: CommandLine): ListArguments {
    /*
     * No non-flag parameters allowed
     * optional --blocked flag
     * optional --completed flag
     * optional --all flag
     */
    ParserUtil.assertNoExtraArgs(commandLine)
    val isAllSet = ParserUtil.getFlagPresence(commandLine, ListCommand.ALL_OPTION.value())
    val isBlockedSet = isAllSet || ParserUtil.getFlagPresence(commandLine, ListCommand.BLOCKED_OPTION.value())
    val isCompletedSet = isAllSet || ParserUtil.getFlagPresence(commandLine, ListCommand.COMPLETED_OPTION.value())
    val isStartedSet = ParserUtil.getFlagPresence(commandLine, ListCommand.STARTED_OPTION.value())
    val isUnblockedSet = (isAllSet
        || ParserUtil.getFlagPresence(commandLine, ListCommand.UNBLOCKED_OPTION.value())
        || !isBlockedSet && !isCompletedSet)
    return ListArguments(isUnblockedSet, isBlockedSet, isCompletedSet, isStartedSet)
  }
}